src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp

```diff
@@ -395,6 +400,211 @@ void reduce_input(ngraph::pass::EinsumDecomposition *einsum_decompose_ptr,
+            offset += alpha;
+        }
+    }
+    const auto identity = ngraph::opset7::Constant::create(input_node.get_element_type(), identity_shape, identity_values);
+    subgraph_nodes.insert(subgraph_nodes.end(), {identity});
+    return subgraph_nodes.back();
+}
+
+ngraph::Output<ngraph::Node> build_multi_identity(ngraph::pass::EinsumDecomposition* einsum_decompose_ptr,
+                                                  const ngraph::Output<ngraph::Node>& input_node,
+                                                  const std::vector<std::string>& repeated_labels,
+                                                  LabelDimMap& label_dim_map,
+                                                  ngraph::NodeVector& subgraph_nodes) {
+    const auto input_shape = input_node.get_shape();
+
+    // initially set multi-identity with identity for the first repeated label
+    NGRAPH_CHECK(repeated_labels.size() > 0);
+    const auto first_repeated_label = repeated_labels[0];
+    NGRAPH_CHECK(label_dim_map.find(first_repeated_label) != label_dim_map.end());
+    auto repeated_label_dims = label_dim_map[first_repeated_label];
+    const auto multi_identity = build_identity(input_node, repeated_label_dims, subgraph_nodes);
+
+    for (size_t label_ind = 1; label_ind < repeated_labels.size(); ++label_ind) {
+        NGRAPH_CHECK(label_dim_map.find(repeated_labels[label_ind]) != label_dim_map.end());
+        repeated_label_dims = label_dim_map[repeated_labels[label_ind]];
+        const auto identity = build_identity(input_node, repeated_label_dims, subgraph_nodes);
+
+        const auto mul = std::make_shared<ngraph::opset7::Multiply>(multi_identity, identity, ngraph::op::AutoBroadcastType::NUMPY);
+        subgraph_nodes.insert(subgraph_nodes.end(), {mul});
+    }
+    return subgraph_nodes.back();
+}
+
+void extract_diagonal(ngraph::pass::EinsumDecomposition* einsum_decompose_ptr,
+                      ngraph::OutputVector& inputs,
+                      std::vector<std::string>& input_subscripts,
+                      size_t input_ind,
+                      ngraph::NodeVector& subgraph_nodes) {
+    // perform sanity check for arguments
+    const auto num_inputs = inputs.size();
+    NGRAPH_CHECK(num_inputs == input_subscripts.size(), "Each input must have own subscript.");
+    NGRAPH_CHECK(input_ind < num_inputs, "Input index is out of range.");
+
+    const auto& input_node = inputs[input_ind];
+    const auto& input_subscript = input_subscripts[input_ind];
+    const auto input_shape = input_node.get_shape();
+
+    std::string resultant_subscript = "";
+    static const std::string ellipsis = "...";
+    const auto labels = ngraph::opset7::Einsum::extract_labels(input_subscript);
+    auto label_dim_map = compute_label_dim_map(input_node.get_partial_shape().rank(), input_subscript);
+    std::vector<std::string> repeated_labels;
+    ngraph::Shape result_shape;
+    ngraph::AxisSet reduced_axes;
+    for (const auto& label : labels) {
```

Comment:
can this loop at least be moved to separate method to make code little bit more readable?

Reply:
Done.